File: src/Javascript/array/74.搜索二维矩阵.js
/*
 * @lc app=leetcode.cn id=74 lang=javascript
 *
 * [74] 搜索二维矩阵
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let left = 0, right = matrix.length * matrix[0].length - 1;
    // binary search
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        // convert number to index
        const midVal = matrix[Math.floor(mid / matrix[0].length)][Math.floor(mid % matrix[0].length)];
        if (midVal === target) return true;
        else if (midVal > target) right = mid - 1;
        else left = mid + 1;
    }
    return false;
};
// @lc code=end

File: src/Java/practice/Second-time/74.搜索二维矩阵.java
/*
 * @lc app=leetcode.cn id=74 lang=java
 *
 * [74] 搜索二维矩阵
 */

// @lc code=start
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int lo = 0, hi = matrix.length * matrix[0].length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int y = mid % matrix[0].length, x = mid / matrix[0].length;
            if (target == matrix[x][y]) return true;
            else if (target < matrix[x][y]) hi = mid - 1;
            else lo = mid + 1;
        }
        return false;
    }
}
// @lc code=end

